044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,345 @@
+For example: `client.quota.callback.static.kafka.admin.security.protocol=SSL`
+
+The strimzi-kafka-operator would be responsible for configuring the admin client to connect to the
+replication listener of the broker.
+
+##### Limit Type Configuration
+
+When operating with cluster sourced volume data the existing limit types applied to the aggregate
+used bytes of all volumes would be meaningless. Nodes enter and exit the active set as part of normal operation.
+So we propose that the existing limiting method should be incompatible with cluster sourced volumes.
+
+Instead, we propose introducing new limit types applied on a per-volume basis. Meaning that there is a single value
+for each limit which we test against each volume. i.e. we do not support limiting based on a specific volume. When
+that limit is exceeded we effectively stop messages being produced.
+
+If the plugin is configured to use a cluster volume source, we also require one limit to be configured.
+
+Defining multiple limit types would be an invalid state.
+
+The limits we want are:
+1. throttle if [availableBytes](#volume) is less-than-or-equal-to some threshold for any volume
+2. throttle if the [availableRatio](#volume) is less-than-or-equal-to some threshold for any volume
+
+For example, to configure a limit when availableBytes is below 1GB:
+- `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow=1000000000`
+
+Another example, to configure a limit when availableRatio is below 0.01 (1%):
+- `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow=0.01`
+
+Defining more than one limit is **invalid**, for example the below is disallowed:
+- `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow=0.05`
+- `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow=5000000000`
+
+#### Throttle Factor Fallback
+
+We are going to depend on using the admin client to get volume information. This brings all the baggage of making a
+connection and dealing with possible failures. Also, we can potentially have an inconsistent view of the world between
+determining the active broker set and asking for a description of all log dirs (because we make two independent API calls)
+. So we need to react somehow to cases where we cannot get the volume data for all active brokers.
+
+Example inconsistent state:
+1. we call `describeCluster` and get a response that says broker 1 and 2 are active
+2. broker 2 shuts down cleanly and is removed from the active set
+3. we call `describeLogDirs( brokerIds = [1,2] )` and only receive descriptions for logdirs on broker 1
+
+We propose introducing a configurable **throttle factor fallback** to be applied in situations where we don't have enough
+information to act. With a default value of 1.0 to optimistically allow all the quota to be used. This would allow
+users to opt in to more pessimistic behaviour like using a factor of 0.0 to prevent writes when we are in an unknown
+state.
```

Comment:
I think we should mention `client.quota.callback.static.throttle.factor.validity-duration` here because it would let people think we'll apply the fallback immediately without completely read the whole proposal.

Reply:
Added a note with link to the other config